src/core/shape_inference/include/proposal_shape_inference.hpp

```diff
@@ -30,78 +32,70 @@ void infer_prop_shape(const OpType* op,
-                              bbox_deltas_ps[1],
-                              ").");
-
-        NODE_VALIDATION_CHECK(op,
-                              class_probs_ps[0].compatible(bbox_deltas_ps[0]),
-                              "Batch size inconsistent between class_probs (",
-                              class_probs_ps[0],
-                              ") and bbox deltas (",
-                              bbox_deltas_ps[0],
-                              ").");
-    }
-
-    if (image_shape_ps.is_static()) {
-        const auto image_shape_elem = image_shape_ps[0].get_length();
-        NODE_VALIDATION_CHECK(op,
-                              image_shape_elem >= 3 && image_shape_elem <= 4,
-                              "Image_shape 1D tensor must have => 3 and <= 4 elements (image_shape_shape[0]",
-                              image_shape_ps[0],
-                              ").");
+    if (image_shape_ps.rank().is_static()) {
+        NODE_VALIDATION_CHECK(
+            op,
+            image_shape_ps.size() == 1 && (image_shape_ps[0].compatible(3) || image_shape_ps[0].compatible(4)),
+            "Image_shape must be 1-D tensor and has got 3 or 4 elements (image_shape_shape[0]",
+            image_shape_ps,
+            ").");
     }
 
-    auto out_dim = DimType{};
-
-    if (class_probs_ps.rank().is_static() && bbox_deltas_ps.rank().is_static()) {
-        OPENVINO_ASSERT(DimType::merge(out_dim, class_probs_ps[0], bbox_deltas_ps[0]));
-    } else if (class_probs_ps.rank().is_static()) {
-        out_dim = class_probs_ps[0];
-    } else if (bbox_deltas_ps.rank().is_static()) {
-        out_dim = bbox_deltas_ps[0];
+    TShape proposed_boxes_shape;
+    proposed_boxes_shape.reserve(2);
+
+    if (class_probs_ps.rank().is_static()) {
+        proposed_boxes_shape.push_back(class_probs_ps[0]);
+
+        if (bbox_deltas_ps.rank().is_static()) {
+            // check anchor count and batch number consistency
+            NODE_VALIDATION_CHECK(op,
+                                  bbox_deltas_ps[1].compatible(class_probs_ps[1] * 2),
+                                  "Anchor number inconsistent between class_probs (",
+                                  class_probs_ps[1] * 2,
+                                  "), and bbox_deltas (",
+                                  bbox_deltas_ps[1],
+                                  ").");
+
+            NODE_VALIDATION_CHECK(op,
+                                  TDim::merge(proposed_boxes_shape[0], proposed_boxes_shape[0], bbox_deltas_ps[0]),
+                                  "Batch size inconsistent between class_probs (",
+                                  class_probs_ps[0],
+                                  ") and bbox deltas (",
+                                  bbox_deltas_ps[0],
+                                  ").");
+        }
     } else {
-        out_dim = Dimension::dynamic();
+        proposed_boxes_shape.emplace_back(ov::util::dim::inf_bound);
```

Comment:
Looks like `batch_size` dimension can be lost if `class_probs_ps.rank()` is dynamic, 
but `bbox_deltas_ps.rank()` static.

Reply:
Good catch will fix.